lib/Echidna/Solidity.hs

```diff
@@ -87,21 +88,44 @@ contracts :: (MonadIO m, MonadThrow m, MonadReader x m, Has SolConf x) => FilePa
 contracts fp = do
   a <- view (hasLens . solcArgs)
   q <- view (hasLens . quiet)
-  pure (a, q) >>= liftIO . solc >>= (\case
+  ls <- view (hasLens . solcLibs)
+  pure (a, q, ls) >>= liftIO . solc >>= (\case
     Nothing -> throwM CompileFailure
     Just m  -> pure . toList $ fst m) where
       usual = ["--combined-json=bin-runtime,bin,srcmap,srcmap-runtime,abi,ast", fp]
-      solc (a, q) = do
+      solc (a, q, ls) = do
         stderr <- if q then UseHandle <$> openFile "/dev/null" WriteMode
                        else pure Inherit
         readSolc =<< writeSystemTempFile ""
-                 =<< readCreateProcess (proc "solc" $ usual <> words a) {std_err = stderr} ""
+                 =<< readCreateProcess (proc "solc" $ usual <> words (a ++ linkLibraries ls)) {std_err = stderr} ""
 
 populateAddresses :: [Addr] -> Integer -> VM -> VM
 populateAddresses []     _ vm = vm
 populateAddresses (a:as) b vm = populateAddresses as b (vm & set (env . EVM.contracts . at a) (Just account))
   where account = initialContract (RuntimeCode mempty) & set nonce 1 & set balance (w256 $ fromInteger b)
 
+-- | Address to load the first library
+addrLibrary :: Addr
+addrLibrary = 0xff
+
+ -- | Load a list of solidity contracts as libraries
+loadLibraries :: (MonadIO m, MonadThrow m, MonadReader x m, Has SolConf x)
+              => [SolcContract] -> Addr -> Addr -> VM -> m VM
+loadLibraries []     _    _ vm = return vm
+loadLibraries (l:ls) la d vm = do vm' <- execStateT (execTx $ Tx (Right bc) d la 0) vm 
+                                  loadLibraries ls (la+1) d vm'
+                               where bc = l ^. creationCode
+
+
+ -- | Generate a string to use as argument in solc to link libraries starting from addrLibrary
+linkLibraries :: [String] -> String
+linkLibraries [] = ""
+linkLibraries ls = "--libraries " ++ linkLibraries' ls addrLibrary
```

Comment:
`= "--libraries " ++ imap (\i x -> concat [x, ":", show $ addrLibrary + i, ","]) ls` might be cleaner than exporting another function

Reply:
Done.